README.md
##PiCute

[Picute](http://picute.mitako.eu/) runs QT on the RaspberryPI,
on top of [pipaOS](http://pipaos.mitako.eu/) Jessie XGUI version, or your preferred Debian based distro.

This project builds the image that will boot on the RPi. It can run QT5 apps on EGLFS, LinuxFB and XCB backends.

It can also be used as a sysroot to cross compile QT5 apps from a Intel computer.

This project is a complete rework of the previous [build system](https://github.com/skarbat/qt5-picute).

###Prepare the host

You will need a Intel 32/64 bit computer with Debian Jessie, and the pieces below need be put on top.

####A cross compiler

The core tools to build the QT5 framework and cross compile your apps.

 * sudo git clone https://github.com/raspberrypi/tools.git /opt/rpi-tools
 * Add one of the following to your PATH (32bit or 64bit host)
    * /opt/rpi-tools/arm-bcm2708/gcc-linaro-arm-linux-gnueabihf-raspbian/bin
    * /opt/rpi-tools/arm-bcm2708/gcc-linaro-arm-linux-gnueabihf-raspbian-x64/bin
 * The default build chain: sudo apt-get install build-essential pkg-config
 * Wayland-scanner: sudo apt-get install libwayland-dev
 * For webengine: sudo apt-get install gperf bison ruby (not entirely sure about ruby)

No extra development libraries needed because they will be pulled from the sysroot.

####The xsysroot tool

[Xsysroot](http://xsysroot.mitako.eu/) simplifies access to the RaspberryPI sysroot image.

 * sudo curl -L "https://raw.githubusercontent.com/skarbat/xsysroot/master/xsysroot" -o /usr/local/bin/xsysroot
 * sudo chmod +x /usr/local/bin/xsysroot
 * sudo xsysroot -U

Call `xsysroot --tools` and install missing software, except for the virtual displays which are not needed.

Finally, make sure you are a sudoer and freedom to do everything as root (ALL=NOPASSWD: ALL)

###Prepare the image

Now get the pipaOS image ready, the so called sysroot.

```
$ mkdir ~/osimages ~/xtmp
$ cd ~/osimages && wget http://pipaos.mitako.eu/download/pipaos-lulo-xgui-4.4.img.gz
$ gunzip pipaos-lulo-xgui-4.4.img.gz
```

Copy the file `xsysroot.conf` to your home directory, then call `xsysroot -p picute -s`.

###Build

Picute.py will take hands, install QT5 dependencies on the sysroot and cross-build QT5 on it:

```
 $ ./buildme
```

All QT5 source code will be cloned into your `~/xtmp` folder, which sits on the host.
Once the build is complete, you will not want this anymore.

###Webengine

Once the buildme script has completed QT5, you can now build the Webengine.

```
$ python -u piwebengine.py picute > webengine.log 2>&1
```

Webengine build takes a long time. On successfeul completion it will have been installed inside the image.

###Cross compile your apps

On your Intel host computer, you'll need to mount the Picute image with `xsysroot -m`.
Then, add `/tmp/picute/usr/local/qt5.5/bin` to your path.

Now from your QT5 project sources, simply call `qmake` followed by `make`,
and it should build the app ready to run on Picute.

An alternative to having the Picute image on your Intel computer,
is to boot Picute on a networked RaspberryPI, and remote mount the sysroot, like this:

`
 $ mkdir /tmp/picute && sudo sshfs -o allow_other sysop@picute-ip-address:/ /tmp/picute
`

Pass `posys` for the network password.

##Webengine 5.7.1 dependencies

The following webengine dependencies can be installed manually:

 * http://httpredir.debian.org/debian/pool/main/f/ffmpeg/libavutil55_3.2-2~bpo8+2_armhf.deb
 * http://httpredir.debian.org/debian/pool/main/f/ffmpeg/libavcodec57_3.2-2~bpo8+2_armhf.deb
 * http://httpredir.debian.org/debian/pool/main/f/ffmpeg/libavformat57_3.2-2~bpo8+2_armhf.deb
 * http://httpredir.debian.org/debian/pool/main/libv/libvpx/libvpx4_1.6.0-2~bpo8+1_armhf.deb

###References

 * https://github.com/CalumJEadie/part-ii-individual-project-dev/wiki/Project-Proposal-Research
 * https://wiki.qt.io/Building_Qt_5_from_Git
 * http://doc.qt.io/qt-5/embedded-linux.html#configuring-for-a-specific-device
 * http://wiki.qt.io/RaspberryPi2EGLFS
 * https://code.qt.io/cgit/qt
 * https://github.com/raspberrypi/tools.git
 * https://www.raspberrypi.org/documentation/linux/kernel/building.md
 * https://www.raspberrypi.org/blog/qt5-and-the-raspberry-pi/
 * http://www.ics.com/blog/building-qt-and-qtwayland-raspberry-pi
 * http://www.intestinate.com/pilfs/beyond.html#wayland
 * https://wiki.merproject.org/wiki/Community_Workspace/RaspberryPi
 * https://github.com/jorgen/yat
 * https://www.youtube.com/watch?v=AtYmJaqxuL4
 * https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=qpi2
 * https://github.com/qtproject/qtwayland
 * http://www.chaosreigns.com/wayland/weston/
 * https://info-beamer.com/blog/raspberry-pi-hardware-video-scaler
 * https://forum.qt.io/topic/48223/webengine-raspberry-pi/2

Albert Casals, February 2016.


changelog

        November 23, 2016: version 1.10
	Upgrade to QT 5.7, with Webengine

	March 20, 2016: version 1.07
	Added webview

	February 30, 2016: version 1.06
	Added qtwayland and compositor support

	February 25, 2016: version 1.05
	QT5.5



piwebengine.py
#!/usr/bin/env python
#
#  Build QT5 webengine
#

import os
import sys
import xsysroot

if __name__ == '__main__':

    # We need a xsysroot profile with QT5 built in it
    if len(sys.argv) > 1:
        xprofile=sys.argv[1]
    else:
        print 'Need a xsysroot profile'
        sys.exit(1)

    # Find and activate the xsysroot profile
    print '>>> Opening xsysroot profile: {}'.format(xprofile)
    try:
        picute=xsysroot.XSysroot(profile=xprofile)
    except:
        print 'You need to create a Xsysroot Picute profile'
        print 'Please see the README file'
        sys.exit(1)

    # Locate Webengine source code directory
    webengine_path=os.path.join(picute.query('tmp'), 'qt5/qtwebengine')

    # Apply temporary patch to build QT5.7 Webengine for the RPI
    # https://bugreports.qt.io/browse/QTBUG-57037
    if not os.path.isdir(webengine_path):
           print '>>> Could not find Webengine path: {}'.format(webengine_path)
           sys.exit(1)
    else:
        patch_file='gyp_run.pro'
        print '>>> Overwriting webengine qmake file: {}'.format(patch_file)
        rc=os.system('cp {} {}/src/core'.format(patch_file, webengine_path))
        if rc:
            print 'Could not apply patch'
            sys.exit(1)
        else:
            print '>>> Webengine patch has been applied'

    # Now mount image if needed
    print '>>> Accessing image...'
    if not picute.is_mounted():
        if not picute.mount():
            sys.exit(1)

    # Step 1: QMAKE
    print '>>> Running Qmake...'
    cmdline_prefix='export PKG_CONFIG_PATH={}/usr/lib/arm-linux-gnueabihf/pkgconfig'.format(picute.query('sysroot'))
    print '>>> cmdline_prefix: ', cmdline_prefix
    rc=os.system('{} ; cd {} ; {}/usr/local/qt5/bin/qmake ' \
                 'WEBENGINE_CONFIG+=use_proprietary_codecs'.format(
                     cmdline_prefix, webengine_path, picute.query('sysroot')))
    if rc:
        print '>>> Qmake failed rc={} :-('.format(rc)
        sys.exit(1)

    # Step 2: MAKE
    print '>>> Running Make...'
    rc=os.system('{} ; cd {} ; make'.format(cmdline_prefix, webengine_path))
    if rc:
        print '>>> Make failed rc={} :-('.format(rc)
        sys.exit(1)

    # Step 3: INSTALL
    print '>>> Running Make Install...'
    rc=os.system('cd {} ; sudo make install'.format(webengine_path))
    if rc:
        print '>>> Make install failed rc={} :-('.format(rc)
        sys.exit(1)

    print '>>> Webengine built and installed'

    # Webengine build complete: Unmount image
    if not picute.umount():
        print '>>> WARNING: Image is busy, most likely installation left some running processes.'
        sys.exit(1)

    sys.exit(0)


native-build.sh
#!/bin/bash
#
# Script to build QT5 natively on the sysroot.
#
# This allows to build QT5 apps on the RaspberryPI
#

qt5_install=$1

qt5_builddeps="libc6-dev libxcb1-dev libxcb-icccm4-dev libxcb-xfixes0-dev \
libxcb-image0-dev libxcb-keysyms1-dev libxcomposite-dev \
libxcb-sync0-dev libxcb-randr0-dev libx11-xcb-dev libxcb-render-util0-dev \
libxrender-dev libxext-dev libxcb-glx0-dev pkg-config \
libssl-dev libraspberrypi-dev libfreetype6-dev libxi-dev libcap-dev \
libwayland-dev libxkbcommon-dev build-essential git-core libfontconfig1-dev \
libasound2-dev libinput-dev libmtdev-dev libproxy-dev libdirectfb-dev \
libts-dev libudev-dev libxcb-xinerama0-dev \
libdbus-1-dev libicu-dev libglib2.0-dev time"

if [ "$qt5_install" == "" ]; then
    echo "Syntax error: native-build.sh <qt5 install directory>"
    exit 1
fi

if [ `uname -m` != "armv7l" ]; then
    echo "Not running in emulated ARM mode - aborting"
    exit 1
fi

apt-get update
apt-get install -y --no-install-recommends $qt5_builddeps

echo ">>> Configure..."
configure_options="
-opengl es2 -eglfs -xcb -device linux-rasp-pi2-g++
-device-option CROSS_COMPILE=/usr/bin/
-sysroot / -opensource -confirm-license -release -skip qtwayland
-prefix /usr/local/qt5 -no-pch -alsa -no-use-gold-linker -qt-xkbcommon
-xkb-config-root /usr/share/X11/xkb -skip qtwebengine -skip qtwebview
-nomake tests -nomake examples -verbose -no-warnings-are-errors
-no-qml-debug -optimized-qmake -strip -fontconfig
"
cd /tmp/qt5
/usr/bin/time ./configure $configure_options

exit 0

echo ">> Make..."
make -j 9

echo ">> Make Install..."
make install


debianize-qt5.py
#!/usr/bin/env python
#
#  debianize-qt5.py
#
#  A rather rustic script to create debian packages from built QT5 libraries, using "dpkg-deb".
#
#  Syntax: qt5-debianize <sysroot directory> <qt5 install path>
#
#  If you run the script inside the chroot, simply pass "/" as the first argument.
#

import sys
import os
import shutil
import glob

# This is the version of the Debian package
qt5_version = '5.7-1'

# This is Debian control file in a skeleton reusable block
control_skeleton='''
Maintainer: Albert Casals <skarbat@gmail.com>
Section: others
Package: {pkg_name}
Version: {pkg_version}
Architecture: armhf
Depends: debconf (>= 0.5.00), {pkg_depends}
Priority: optional
Description: {pkg_description}

'''

# These are the X11 libraries on which QT5 depends on, provides XCB subsystem
extra_deps = 'libx11-xcb1, libxcb-icccm4, libxcb-xfixes0, libxcb-image0, libxcb-keysyms1, libxcomposite1, ' \
    'libxcb-randr0, libxcb-render-util0, libxrender1, libxext6, libxcb-glx0, libfontconfig1, ' \
    'libinput5, libts-0.0-0, libjpeg62-turbo, libasound2, libproxy1, libicu52'

# These are the packages we are building
# For the moment we are collecting everyting in one single Debian pkg
packages=[

    # TODO: Split all the libraries into individul packages (core, network, gui, etc)
    # There are 36 libraries and 17 plugin folders on QT 5.5

    { 'fileset': [ 'imports',
                   'lib/lib*.so.*',
                   'lib/lib*.so',
                   'lib/fonts/*',
                   'plugins',
                   'qml'
                   ],

      'pkg_name': 'libqt5all',
      'pkg_version': qt5_version,
      'pkg_depends': '{}, libraspberrypi0'.format(extra_deps),
      'pkg_description': 'All QT5 Libraries and basic tools' },

    { 'fileset': [ 'bin', 'include', 'mkspecs', # FIXME: mkspecs has large number of devices we will never use
                   'lib/*.a', 'lib/*.la', 'lib/*.prl', 'lib/cmake', 'lib/pkgconfig', 'translations' ],
      'pkg_name': 'libqt5all-dev',
      'pkg_version': qt5_version,
      'pkg_depends': 'libqt5all, libraspberrypi-dev',
      'pkg_description': 'All QT5 Development files' }
]


if __name__ == '__main__':

    if len(sys.argv) < 2:
        print 'Syntax: native-debianize <sysroot directory> <qt directory>'
        print '        sysroot : pathname of the image mount point'
        print '        qt directory : pathname inside the root (i.e. /usr/local/qt5)'
        sys.exit(1)
    else:
        root_directory=sys.argv[1]
        source_directory=sys.argv[2]
        complete_source='{}/{}'.format(root_directory, source_directory)

    # Sanity check
    if not os.path.exists(complete_source):
        print 'error: path not found', complete_source
        sys.exit(1)

    for pkg in packages:

        # allocate a versioned directory name for the package
        versioned_pkg_name = 'pkgs/{}_{}'.format(pkg['pkg_name'], qt5_version)
        print 'Processing package {}...'.format(versioned_pkg_name)

        # extract the files from the root file system preparing them for packaging
        target_directory = '{}/{}'.format (versioned_pkg_name, source_directory)

        for files in pkg['fileset']:

            # Complete the pathname to the target directory
            last_path = os.path.dirname(files)
            target_files_path='{}/{}'.format(target_directory, last_path)

            print 'Extracting {} into {}...'.format(os.path.join(complete_source, files), target_files_path)
            if not os.path.exists(target_files_path):
                os.makedirs(target_files_path)

            os.system('cp -rvP {} {}'.format(os.path.join(complete_source, files), target_files_path))

        # Cleanup anything related to webengine
        cmd='find ' + target_files_path + ' -iname \*webengine\* -exec rm -rfv {} \;'
        os.system(cmd)

        # create the Debian control file for "dpkg-deb" tool to know what to pack
        debian_dir=os.path.join(versioned_pkg_name, 'DEBIAN')
        if not os.path.exists(debian_dir):
            os.makedirs(debian_dir)
        with open(os.path.join(debian_dir, 'control'), 'w') as control_file:
            control_file.writelines(control_skeleton.format(**pkg))

        # copy the shlibs file for the runtime package
        if pkg['pkg_name'] == 'libqt5all':
            os.system('mkdir -p {}/../var/lib/dpkg/info'.format(debian_dir))
            os.system('cp -v {} {}/../var/lib/dpkg/info/libqt5all.shlibs'.format('shlibs.local-qt5', debian_dir))

        # finally call dpkg-deb and generate a debian package
        rc=os.system('dpkg-deb --build {}'.format(versioned_pkg_name))
        if not rc:
            print 'Package {} created correctly'.format(versioned_pkg_name)
        else:
            print 'WARNING: Error creating package {}'.format(versioned_pkg_name)


debianize-webengine.py
#!/usr/bin/env python
#
#  debianize-webengine.py
#
#  A rather rustic script to package QT5 Webengine into a Debian package
#
#  Syntax: webengine-debianize <sysroot directory> <qt5 install path>
#
#  If you run the script inside the chroot, simply pass "/" as the first argument.
#

import sys
import os
import shutil
import glob

# This is the version of the Debian package
qt5_version = '5.7-1'

# This is Debian control file in a skeleton reusable block
control_skeleton='''
Maintainer: Albert Casals <skarbat@gmail.com>
Section: others
Package: {pkg_name}
Version: {pkg_version}
Architecture: armhf
Depends: debconf (>= 0.5.00), {pkg_depends}
Priority: optional
Description: {pkg_description}

'''

# Webengine extra core dependencies
extra_deps=''

packages=[

    # QtWebEngine Runtime package
    { 'fileset': [ 'translations/qtwebengine_locales/*',
                   'resources/qtwebengine_resources*pak',
                   'resources/qtwebengine_devtools_resources.pak',
                   'resources/icudtl.dat',
                   'lib/libQt5WebEngine*',
                   'qml/QtWebEngine/*',
                   'libexec/QtWebEngineProcess'
               ],
  
      'pkg_name': 'libqt5webengine',
      'pkg_version': qt5_version,
      'pkg_depends': 'libqt5all{}'.format(extra_deps),
      'pkg_description': 'QT5 WebEngine Libraries and basic tools'
  },

    # QtWebEngine Developer's package
    { 'fileset': [ 'include/QtWebEngineWidgets/*',
                   'include/QtWebEngine/*',
                   'include/QtWebEngineCore/*',
                   'mkspecs/modules/qt_lib_webengine*',
                   'lib/cmake/Qt5WebEngine*',
                   'lib/pkgconfig/Qt5WebEngine*'
               ],
  
      'pkg_name': 'libqt5webengine-dev',
      'pkg_version': qt5_version,
      'pkg_depends': 'libqt5webengine',
      'pkg_description': 'QT5 WebEngine Libraries and basic tools'
  }
    
]


if __name__ == '__main__':

    if len(sys.argv) < 2:
        print 'Syntax: webengine-debianize <sysroot directory> <qt directory>'
        print '        sysroot : pathname of the image mount point'
        print '        qt directory : pathname inside the root (i.e. /usr/local/qt-v5.4.1)'
        sys.exit(1)
    else:
        root_directory=sys.argv[1]
        source_directory=sys.argv[2]
        complete_source='{}/{}'.format(root_directory, source_directory)

    # Sanity check
    if not os.path.exists(complete_source):
        print 'error: path not found', complete_source
        sys.exit(1)

    for pkg in packages:

        # allocate a versioned directory name for the package
        versioned_pkg_name = 'pkgs/{}_{}'.format(pkg['pkg_name'], qt5_version)
        print 'Processing package {}...'.format(versioned_pkg_name)

        # extract the files from the root file system preparing them for packaging
        target_directory = '{}/{}'.format (versioned_pkg_name, source_directory)

        for files in pkg['fileset']:

            # Complete the pathname to the target directory
            last_path = os.path.dirname(files)
            target_files_path='{}/{}'.format(target_directory, last_path)

            print 'Extracting {} into {}...'.format(os.path.join(complete_source, files), target_files_path)
            if not os.path.exists(target_files_path):
                os.makedirs(target_files_path)

            os.system('cp -rvP {} {}'.format(os.path.join(complete_source, files), target_files_path))

        # create the Debian control file for "dpkg-deb" tool to know what to pack
        debian_dir=os.path.join(versioned_pkg_name, 'DEBIAN')
        if not os.path.exists(debian_dir):
            os.makedirs(debian_dir)
        with open(os.path.join(debian_dir, 'control'), 'w') as control_file:
            control_file.writelines(control_skeleton.format(**pkg))

        # copy the shlibs file for the runtime package
        if pkg['pkg_name'] == 'libqt5webengine':
            os.system('mkdir -p {}/../var/lib/dpkg/info'.format(debian_dir))
            os.system('cp -v {} {}/../var/lib/dpkg/info/libqt5webengine.shlibs'.format('shlibs.local-webengine', debian_dir))

        # finally call dpkg-deb and generate a debian package
        rc=os.system('dpkg-deb --build {}'.format(versioned_pkg_name))
        if not rc:
            print 'Package {} created correctly'.format(versioned_pkg_name)
        else:
            print 'WARNING: Error creating package {}'.format(versioned_pkg_name)


picute.py
#!/usr/bin/env python
#
#  The MIT License (MIT)
#
#  Copyright (c) 2016 Albert Casals - albert@mitako.eu
#
#  Permission is hereby granted, free of charge, to any person obtaining a copy
#  of this software and associated documentation files (the "Software"), to deal
#  in the Software without restriction, including without limitation the rights
#  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
#  copies of the Software, and to permit persons to whom the Software is
#  furnished to do so, subject to the following conditions:
#
#  The above copyright notice and this permission notice shall be included in
#  all copies or substantial portions of the Software.
#
#  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
#  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
#  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
#  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
#  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
#  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
#  THE SOFTWARE.
#
#  picute.py
#
#  Program that builds QT5.5 on pipaOS 4.4 XGUI version.
#
#  See the README for details.
#

import os
import sys
import time
import platform
import multiprocessing

import xsysroot

# Release version
__version__='1.10'


def test_image(picute, install_path):
    '''
    Performs minimalist tests on the built image
    '''
    failed=0
    sysroot=picute.query('sysroot')
    qt5_dir='{}/{}'.format(sysroot, install_path)
    qt5_qmake=os.path.join(qt5_dir, 'bin/qmake')
    qt5_wayland_dir=os.path.join(qt5_dir, 'include/QtWaylandClient')
    qt5_wayland_so=os.path.join(qt5_dir, 'lib/libQt5WaylandClient.so.5')
    wayland_scanner=os.path.join(qt5_dir, 'bin/qtwaylandscanner')

    if not os.path.isdir(qt5_dir):
        failed+=1
        print 'FAILED: qt5 dir not found: {}'.format(qt5_dir)
    if not os.path.isfile(qt5_qmake):
        failed+=1
        print 'FAILED: qmake not found: {}'.format(qt5_dir)
    if not os.path.isdir(qt5_wayland_dir):
        failed+=1
        print 'FAILED: qtwayland lib directory: {}'.format(qt5_wayland_dir)
    if not os.path.isfile(qt5_wayland_so):
        failed+=1
        print 'FAILED: could not find QtWayland .so: {}'.format(qt5_wayland_so)
    if not os.path.isfile(wayland_scanner):
        failed+=1
        print 'FAILED: could not find Qt Wayland scanner: {}'.format(wayland_scanner)

    return failed


def baptize_image(picute):
    '''
    Prepare the image for the first time. yay!!
    '''

    # make sure the image is not currently in use
    if picute.is_mounted():
        if not picute.umount():
            return False

    # renew the image so we start from clean
    if not picute.renew():
        return False
    else:
        # once renewed, expand it to grow in size, qt5 wouldn't fit
        picute.umount()
        if not picute.expand():
            print 'error expanding image size to {}'.format(picute.query('qcow_size'))
            return False
        else:
            picute.mount()

    # baptize the picute version
    picute.edfile('/etc/picute_version', 'picute v{} - {}'.format(__version__, time.ctime()))

    # set the system hostname
    picute_hostname='picute'
    picute_hosts_file='/etc/hosts'
    picute.edfile('/etc/hostname', picute_hostname)
    picute.edfile(picute_hosts_file, '127.0.0.1 localhost')
    picute.edfile(picute_hosts_file, '127.0.0.1 {}'.format(picute_hostname), append=True)

    # custom firmware config.txt settings
    src_config_txt='config.txt'
    dst_config_txt=os.path.join(picute.query('sysboot'), src_config_txt)
    rc=os.system('sudo cp -fv {} {}'.format(src_config_txt, dst_config_txt))
    if rc:
        print 'WARNING: could not copy config.txt rc={}'.format(rc)

    # custom kernel boot parameters
    src_cmdline='cmdline.txt'
    dst_cmdline=os.path.join(picute.query('sysboot'), src_cmdline)
    rc=os.system('sudo cp -fv {} {}'.format(src_cmdline, dst_cmdline))
    if rc:
        print 'WARNING: could not copy cmdline rc={}'.format(rc)

    # Put the system up to date and install QT5 build dependencies
    # We might need more, for example TLS and further backends
    qt5_builddeps='libc6-dev libxcb1-dev libxcb-icccm4-dev libxcb-xfixes0-dev ' \
        'libxcb-image0-dev libxcb-keysyms1-dev libxcomposite-dev ' \
        'libxcb-sync0-dev libxcb-randr0-dev libx11-xcb-dev libxcb-render-util0-dev ' \
        'libxrender-dev libxext-dev libxcb-glx0-dev pkg-config ' \
        'libssl-dev libraspberrypi-dev libfreetype6-dev libxi-dev libcap-dev ' \
        'libwayland-dev libxkbcommon-dev build-essential git-core libfontconfig1-dev ' \
        'libasound2-dev libinput-dev libmtdev-dev libproxy-dev libdirectfb-dev ' \
        'libts-dev libudev-dev libxcb-xinerama0-dev ' \
        'libdbus-1-dev libicu-dev libglib2.0-dev '

        # was needed for ffmpeg native libs
        #'libavutil-dev=7:3.2-2~bpo8+2 libavcodec-dev=7:3.2-2~bpo8+2 libavformat-dev=7:3.2-2~bpo8+2 ' \
        #'libvpx=1.6.0-2~bpo8+1 libopus-dev libopusfile-dev libwebp-dev '
    
    # Webengine specific dependencies (for FFMPEG native build mode) are pulled from Debian ARM backports repo
    #picute.edfile('/etc/apt/sources.list.d/raspberrypi-backports.list',
    #              'deb http://httpredir.debian.org/debian jessie-backports main contrib')
    #picute.execute('apt-key adv --keyserver keyserver.ubuntu.com --recv-keys 8B48AD6246925553')
    #picute.execute('apt-key adv --keyserver keyserver.ubuntu.com --recv-keys 7638D0442B90D010')

    if picute.execute('apt-get update'):
        return False
    if picute.execute('apt-get install -y --no-install-recommends {}'.format(qt5_builddeps)):
        return False

    # Fix relative symlinks to libdl and libm (so called fixQualifiedPaths in QT jargon)
    # TODO: Use readlink instead of hardcoded destination to .so versioned filename,
    #       but no big deal really. This is a development / builder box.
    picute.execute('rm -fv /usr/lib/arm-linux-gnueabihf/libdl.so')
    picute.execute('cp -fv /lib/arm-linux-gnueabihf/libdl.so.2 /usr/lib/arm-linux-gnueabihf/libdl.so')
    picute.execute('rm -fv /usr/lib/arm-linux-gnueabihf/libm.so')
    picute.execute('cp -fv /lib/arm-linux-gnueabihf/libm.so.6 /usr/lib/arm-linux-gnueabihf/libm.so')

    # Fix relative path for libudev library. This fixes plugging HID devices on-the-fly
    # The reason for this ugly patch is that "configure" seems to have lost the relative sysroot directory
    picute.execute('rm -fv /usr/lib/arm-linux-gnueabihf/libudev.so')
    os.system ('sudo ln -sfv {}/lib/arm-linux-gnueabihf/libudev.so.1.5.0 ' \
               '{}/usr/lib/arm-linux-gnueabihf/libudev.so'.format(
                   picute.query('sysroot'),
                   picute.query('sysroot')))

    return True



if __name__ == '__main__':

    help_welcome='Welcome to picute version {} builder'.format(__version__)
    help_syntax=' Syntax: picute.py <xsysroot profile> [--baptize]'
    output_image='picute-{}.img'.format(__version__)
    baptize=False
    wayland=False
    convert_image=False

    # Xsysroot profile name that holds the original pipaOS image
    # (See the file xsysroot.conf for details)
    if len(sys.argv) < 2:
        print help_syntax
        sys.exit(1)
    else:
        xsysroot_profile_name=sys.argv[1]

    print help_welcome

    # Do we want to baptize the image from scratch?
    if len(sys.argv) == 3:
        if sys.argv[2] == '--baptize':
            print 'WARNING: Baptizing image from scratch'
            baptize=True
        else:
            print help_syntax
            sys.exit(1)
                    
    # Find and activate the xsysroot profile
    try:
        picute=xsysroot.XSysroot(profile=xsysroot_profile_name)
    except:
        print 'You need to create a Xsysroot Picute profile'
        print 'Please see the README file'
        sys.exit(1)

    # prepare some commonly used paths and variables
    xtmp=picute.query('tmp')
    host_numcpus=multiprocessing.cpu_count()
    qt5_version='5.7.1'
    qt5_srcdir=os.path.join(xtmp, 'qt5')
    rpi_tools='/opt/rpi-tools'
    qt5_path_prefix='/usr/local/qt5'

    xgcc_path32='/arm-bcm2708/gcc-linaro-arm-linux-gnueabihf-raspbian/bin'
    xgcc_path64='/arm-bcm2708/gcc-linaro-arm-linux-gnueabihf-raspbian-x64/bin'

    # find out if 32 or 64 bit host system
    if platform.architecture()[0] == '64bit':
        xgcc_path=xgcc_path64
    else:
        xgcc_path=xgcc_path32
        
    xgplusplus='{}/{}/arm-linux-gnueabihf-g++'.format(rpi_tools, xgcc_path)

    init_repo_done=os.path.join(xtmp, 'init_repo_done.chk')
    configure_done=os.path.join(xtmp, 'configure_done.chk')
    make_done=os.path.join(xtmp, 'make_done.chk')
    make_install_done=os.path.join(xtmp, 'make_install_done.chk')

    configure_opts='-opengl es2 -eglfs -xcb -device linux-rasp-pi2-g++ ' \
        '-device-option CROSS_COMPILE={}/{} ' \
        '-sysroot {} -opensource -confirm-license -release -skip qtwayland ' \
        '-prefix {} -pkg-config -no-pch -alsa -no-use-gold-linker -qt-xkbcommon ' \
        '-xkb-config-root /usr/share/X11/xkb -skip qtwebengine -skip qtwebview ' \
        '-nomake tests -nomake examples -verbose -no-warnings-are-errors ' \
        '-no-qml-debug -optimized-qmake -strip -fontconfig -no-sql-sqlite'.format(
            rpi_tools, xgcc_path + '/arm-linux-gnueabihf-',
            picute.query('sysroot'), qt5_path_prefix)

    # Start from a clean image
    time_start=time.time()
    if baptize:
        if not baptize_image(picute):
            print 'Error baptizing image'
            sys.exit(1)
    else:
        if not picute.is_mounted():
            if not picute.mount():
                sys.exit(1)

    if not os.path.isfile(xgplusplus):
        print '>>> Checking QT5 build software on the host'
        print 'Could not find the g++ cross compiler at {}'.format(xgplusplus)
        sys.exit(0)

    # Step 1: clone the sources
    if not os.path.isdir(qt5_srcdir):
        print '>>> Cloning QT5 sources'
        rc=os.system('git clone --branch {} git://code.qt.io/qt/qt5.git {}'.format(qt5_version, qt5_srcdir))
        if os.WEXITSTATUS(rc):
            print 'Could not clone QT5 sources rc={} :-('.format(rc)
            sys.exit(1)

    # Step 2: init-repository
    if not os.path.isfile(init_repo_done):
        print '>>> Init repository'
        rc=os.system('cd {} && perl init-repository'.format(qt5_srcdir))
        if os.WEXITSTATUS(rc):
            print 'init-repository failed rc={} :-('.format(rc)
            sys.exit(1)
        else:
            os.system('touch {}'.format(init_repo_done))

    # Step 3: configure
    if not os.path.isfile(configure_done) or baptize:
        print '>>> Configure {}'.format(configure_opts)
        rc=os.system('cd {} && ./configure {}'.format(
            qt5_srcdir, configure_opts))
        if os.WEXITSTATUS(rc):
            print 'configure failed rc={} :-('.format(rc)
            sys.exit(1)
        else:
            os.system('touch {}'.format(configure_done))

    # Step 4: make
    if not os.path.isfile(make_done) or baptize:
        print '>>> Make ...'
        rc=os.system('cd {} && make -j {}'.format(qt5_srcdir, host_numcpus))
        if os.WEXITSTATUS(rc):
            print 'make failed rc={} :-('.format(rc)
            sys.exit(1)
        else:
            if wayland:
                # Now make qtwayland with compositor
                print '>>> QtWayland building...'
                rc1=os.system('cd {}/qtwayland && ../qtbase/bin/qmake CONFIG+=wayland-compositor'.format(qt5_srcdir))
                rc2=os.system('cd {}/qtwayland && make'.format(qt5_srcdir))
                rc3=os.system('cd {}/qtwayland && sudo make install'.format(qt5_srcdir))
                print '>>> QtWayland built: rc1={} rc2={} rc3={}'.format(rc1, rc2, rc3)

            os.system('touch {}'.format(make_done))

    # Step 5: make install
    if not os.path.isfile(make_install_done) or baptize:
        print '>>> Make install...'
        rc=os.system('cd {} && sudo make install'.format(qt5_srcdir))
        if os.WEXITSTATUS(rc):
            print 'make install failed :-('
            sys.exit(1)
        else:
            os.system('touch {}'.format(make_install_done))

    # Step 6: run some basic tests on the image
    print '>>> testing image'
    num_failed_tests = test_image(picute, qt5_path_prefix)
    print '>>> tests complete failed={}'.format(num_failed_tests)

    # Create QT5 core library packages
    rc=os.system('python debianize-qt5.py {} {}'.format(picute.query('sysroot'), qt5_path_prefix))

    # unmount the image and release it
    print '>>> releasing and cleaning image'
    if not picute.umount():
        print 'WARNING: Image is busy, most likely installation left some running processes, skipping conversion'
        sys.exit(1)
    else:
        picute.zerofree(verbose=False)

    if convert_image:
        # Convert the xsysroot image to a raw format ready to flash and boot
        qcow_image=picute.query('qcow_image')
        print 'Converting image {}...'.format(qcow_image)
        if os.path.isfile(output_image):
            os.unlink(output_image)

        rc = os.system('qemu-img convert {} {}'.format(qcow_image, output_image))

    time_end=time.time()
    print 'Process finished in {} secs - image ready at {}'.format(time_end - time_start, output_image)
    sys.exit(rc)
